File: readersWritersGo/reader.go
package main

import (
	"sync"

	log "github.com/Sirupsen/logrus"
)

// Reader definition
type Reader struct {
	id       int
	resource *Resource
	wg       *sync.WaitGroup
}

// NewReader creates a Reader
func NewReader(id int, resource *Resource, wg *sync.WaitGroup) *Reader {
	return &Reader{id: id, resource: resource, wg: wg}
}

func (r *Reader) read() {
	defer r.wg.Done()

	log.Info("Reader starting ", r.id)

	r.resource.read(r.id)

	log.Info("Reader done ", r.id)
}

File: readersWritersJava/src/main/java/Reader.java

public class Reader implements Runnable {

    private final int id;
    private final Resource resource;

    public Reader(final int id, final Resource resource) {
        this.id = id;
        this.resource = resource;
    }

    @Override
    public void run() {
        try {
            //System.out.printf("Reader %d starting\n", id);

            resource.read(id);

            //System.out.printf("Reader %d is done\n", id);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }
}